### PR TITLE
Support running benchmarks on netfx

### PR DESCRIPTION
**Pull request by Anipik**
Fixes https://github.com/dotnet/machinelearning/issues/1945

This PR makes changes in order to enable developers to run ML benchmarks against .NetFramework

### CODE REVIEW

File: test/Directory.Build.props [hunk 1]
```diff
@@ -2,10 +2,9 @@
   <Import Project="..\Directory.Build.props" />
 
   <PropertyGroup>
-    <DefaultTestTargetFramework>netcoreapp2.1</DefaultTestTargetFramework>
-    <DefaultTestTargetFramework Condition="$(Configuration.EndsWith('-netfx'))">net462</DefaultTestTargetFramework>
-    <DefaultTestTargetFramework Condition="'$(UseIntrinsics)' == 'true'">netcoreapp3.0</DefaultTestTargetFramework>
-    <TargetFramework>$(DefaultTestTargetFramework)</TargetFramework>
+    <TargetFramework>netcoreapp2.1</TargetFramework>
+    <TargetFramework Condition="$(Configuration.EndsWith('-netfx'))">net462</TargetFramework>
```
Comment (eerhardt) [inline test/Directory.Build.props]: (not part of this change) - do you remember why we choose `net462` instead of `net461`?

File: test/Microsoft.ML.Benchmarks/Harness/Configs.cs [hunk 2]
```diff
@@ -22,6 +25,7 @@ public RecommendedConfig()
 
         protected virtual Job GetJobDefinition()
             => Job.Default
+                .With(msbuildArguments)
```
Comment (eerhardt) [inline test/Microsoft.ML.Benchmarks/Harness/Configs.cs]: Isn't this a duplicate of line 20 above?

```
                .WithCustomBuildConfiguration(GetBuildConfigurationName())
```

Why do we need both?

File: test/Microsoft.ML.Benchmarks/Harness/Configs.cs [hunk 3]
```diff
@@ -55,6 +65,7 @@ public class TrainConfig : RecommendedConfig
     {
         protected override Job GetJobDefinition()
             => Job.Dry // the "Dry" job runs the benchmark exactly once, without any warmup to mimic real-world scenario
+                  .With(msbuildArguments)
```
Comment (eerhardt) [inline test/Microsoft.ML.Benchmarks/Harness/Configs.cs]: Shouldn't this be

```
.WithCustomBuildConfiguration(GetBuildConfigurationName())
```
?

**Comment by eerhardt** [comment]
> using System.Collections.Generic;

Can you add a copyright header while you are in here?

---
Refers to: test/Microsoft.ML.Benchmarks/Harness/Configs.cs:1 in 42b36d7. [](commit_id = 42b36d7e91439554bb7a27afd3ea228a671964c8, deletion_comment = False)

File: test/Directory.Build.props (hunk 1 above)
Comment (Anipik) [reply test/Directory.Build.props]: I remember trying it with net471 and santi suggested to try with net462 to just make the build successfull. It was a while back. and i didnt try building with any other versions

Comment (danmosemsft) [reply test/Directory.Build.props]: If net461 works we should probably use that everywhere, since that supports NET Standard 2.0 it seems like the natural choice.

File: test/Microsoft.ML.Benchmarks/Harness/ProjectGenerator.cs [hunk 4]
```diff
@@ -35,7 +42,7 @@ protected override void GenerateProject(BuildPartition buildPartition, Artifacts
   <PropertyGroup>
     <OutputType>Exe</OutputType>
     <OutputPath>bin\{buildPartition.BuildConfiguration}</OutputPath>
-    <TargetFramework>{TargetFrameworkMoniker}</TargetFramework>
+    <TargetFramework>{TargetFrameworkMoniker}</TargetFramework>{rid}
```
Comment (adamsitnik) [inline test/Microsoft.ML.Benchmarks/Harness/ProjectGenerator.cs]: I think that MsBuild ignores empty `<RuntimeIdentifier></RuntimeIdentifier>` so we can remove the condition and the need for extra variable

```suggestion
    <TargetFramework>{TargetFrameworkMoniker}</TargetFramework>
    <RuntimeIdentifier>{runtimeIdentifier}</RuntimeIdentifier>
```

File: test/Microsoft.ML.Benchmarks/Harness/ProjectGenerator.cs [hunk 5]
```diff
@@ -53,6 +60,9 @@ protected override void GenerateProject(BuildPartition buildPartition, Artifacts
   </ItemGroup>
 </Project>");
 
+        protected override string GetBinariesDirectoryPath(string buildArtifactsDirectoryPath, string configuration)
```
Comment (adamsitnik) [inline test/Microsoft.ML.Benchmarks/Harness/ProjectGenerator.cs]: it would be good to add a comment explaining why we need that

File: test/Microsoft.ML.Benchmarks/Numeric/Ranking.cs [hunk 6]
```diff
@@ -85,8 +85,11 @@ public void SetupScoringSpeedTests()
             if (!File.Exists(_mslrWeb10k_Train))
                 throw new FileNotFoundException(string.Format(Errors.DatasetNotFound, _mslrWeb10k_Train));
 
+#if !NET461
+            _modelPath_MSLR = Path.Combine(Directory.GetParent(typeof(RankingTest).Assembly.CodeBase.Substring(8)).FullName, @"FastTreeRankingModel.zip");
```
Comment (adamsitnik) [inline test/Microsoft.ML.Benchmarks/Numeric/Ranking.cs]: `Substring(8)` is very fragile. Can find a way to avoid the substring? Sth like:

```suggestion
            _modelPath_MSLR = Path.Combine(Path.GetDirectoryName(typeof(RankingTest).Assembly.Location), "FastTreeRankingModel.zip");
```

File: test/Microsoft.ML.Benchmarks/Text/MultiClassClassification.cs [hunk 7]
```diff
@@ -104,7 +104,11 @@ public void SetupScoringSpeedTests()
             if (!File.Exists(_dataPath_Wiki))
                 throw new FileNotFoundException(string.Format(Errors.DatasetNotFound, _dataPath_Wiki));
 
+#if NET461
+            _modelPath_Wiki = Path.Combine(Directory.GetParent(typeof(MultiClassClassificationTest).Assembly.CodeBase.Substring(8)).FullName, @"WikiModel.zip");
```
Comment (adamsitnik) [inline test/Microsoft.ML.Benchmarks/Text/MultiClassClassification.cs]: again: substring

File: test/Microsoft.ML.Benchmarks/Harness/ProjectGenerator.cs [hunk 8]
```diff
@@ -25,8 +25,15 @@ namespace Microsoft.ML.Benchmarks.Harness
     /// </summary>
     public class ProjectGenerator : CsProjGenerator
     {
+        private string rid = string.Empty;
+        private string ridValue = string.Empty;
```
Comment (adamsitnik) [inline test/Microsoft.ML.Benchmarks/Harness/ProjectGenerator.cs]:
```suggestion
```

File: test/Microsoft.ML.Benchmarks/Numeric/Ranking.cs [hunk 9]
```diff
@@ -85,7 +85,7 @@ public void SetupScoringSpeedTests()
             if (!File.Exists(_mslrWeb10k_Train))
                 throw new FileNotFoundException(string.Format(Errors.DatasetNotFound, _mslrWeb10k_Train));
 
-            _modelPath_MSLR = Path.Combine(Directory.GetCurrentDirectory(), @"FastTreeRankingModel.zip");
+            _modelPath_MSLR = Path.Combine(Path.GetDirectoryName(typeof(RankingTest).Assembly.Location), "FastTreeRankingModel.zip");
```
Comment (eerhardt) [inline test/Microsoft.ML.Benchmarks/Numeric/Ranking.cs]: Why is this change necessary?

Comment (Anipik) [reply test/Microsoft.ML.Benchmarks/Numeric/Ranking.cs]: otherwise on .netfx it creates this new model file in the root directory

Comment (eerhardt) [reply test/Microsoft.ML.Benchmarks/Numeric/Ranking.cs]: This confuses me. What do you mean by "the root directory"? Why would the current netfx have its current directory set to that?

also, the code is still using Directory.GetCurrentDirectory() in `Test_Multiclass_WikiDetox_BigramsAndTrichar_OVAAveragedPerceptron`. WHy doesn't that need to change?

Comment (Anipik) [reply test/Microsoft.ML.Benchmarks/Numeric/Ranking.cs]: > Why would the current netfx have its current directory set to that?

I don't have nay concrete explaination for why it is setting current directory to ```C:\git\Mlnet\test\Microsoft.ML.Benchmarks```

Comment (eerhardt) [reply test/Microsoft.ML.Benchmarks/Numeric/Ranking.cs]: @adamsitnik - do you happen to know? Is this something BenchmarkDotNet is controlling?

Comment (adamsitnik) [reply test/Microsoft.ML.Benchmarks/Numeric/Ranking.cs]: @eerhardt  Yes BDN sets the working directory to the folder with built dlls for .NET Core and to null for FX.  The problem is that I dont remember why and I would prefer not to change it if I dont have to ;) It most probably has to deal with the fact that we run executable for FX and do `dotnet name.dll` for Core

https://github.com/dotnet/BenchmarkDotNet/blob/573566b70f0ff27572b9268002fe9748be4de6ea/src/BenchmarkDotNet/Toolchains/DotNetCli/DotNetCliExecutor.cs#L72

https://github.com/dotnet/BenchmarkDotNet/blob/573566b70f0ff27572b9268002fe9748be4de6ea/src/BenchmarkDotNet/Toolchains/Executor.cs#L35

**Comment by Anipik** [comment]
@eerhardt is there anything else that needs to be done here ?

File: test/Microsoft.ML.TestFramework/BaseTestBaseline.cs [hunk 10]
```diff
@@ -34,7 +34,7 @@ public abstract partial class BaseTestBaseline : BaseTestClass
 
         protected BaseTestBaseline(ITestOutputHelper output) : base(output)
         {
-#if NET462
+#if NET461
```
Comment (eerhardt) [inline test/Microsoft.ML.TestFramework/BaseTestBaseline.cs]: NOTE: I just learned about `VersionlessImplicitFrameworkDefine` in the .NET Core SDK - https://github.com/dotnet/sdk/pull/2073. It appears this can be changed to `#if NETFRAMEWORK`

Comment (Anipik) [reply test/Microsoft.ML.TestFramework/BaseTestBaseline.cs]: I will add this change in my next PR
